skbio/io/format/binary_dm.py

```diff
@@ -0,0 +1,178 @@
+"""
+Simple binary dissimilarity matrix format (:mod:`skbio.io.format.binary_dm`)
+============================================================================
+
+.. currentmodule:: skbio.io.format.binary_dm
+
+The Binary DisSimilarity Matrix format (``binary_dm``) encodes a binary
+representation for dissimilarity and distance matrices. The format is
+designed to facilitate rapid random access to individual rows or columns of
+a hollow matrix.
+
+Format Support
+--------------
+**Has Sniffer: Yes**
+
++------+------+---------------------------------------------------------------+
+|Reader|Writer|                          Object Class                         |
++======+======+===============================================================+
+|Yes   |Yes   |:mod:`skbio.stats.distance.DissimilarityMatrix`                |
++------+------+---------------------------------------------------------------+
+|Yes   |Yes   |:mod:`skbio.stats.distance.DistanceMatrix`                     |
++------+------+---------------------------------------------------------------+
+
+Format Specification
+--------------------
+The binary dissimilarity matrix and object identifiers are stored within an
+HDF5[1]_ file. Both datatypes are represented by their own datasets. The
+`ids` dataset is of a variable length unicode type, while the
+`matrix` dataset are floating point. The shape of the `ids` is
+`(N,)`, and the shape of the `dissimilarities` is `(N, N)`. The diagonal of
+`matrix` are all zeros.
+
+The dissimilarity between `ids[i]` and `ids[j]` is interpreted
+to be the value at `matrix[i, j]`. `i` and `j` are integer indices.
+
+Required attributes:
+
++-----------+---------------------+------------------------------+
+|Attribute  |Value                |Description                   |
+|           |type                 |                              |
++===========+=====================+==============================+
+|format     |string               |A string identifying the file |
+|           |                     |as Binary DM format           |
++-----------+---------------------+------------------------------+
+|version    |string               |The version of the current    |
+|           |                     |Binary DM format              |
```

Comment:
How do you feel about the discordance between these being considered "required", but the sniffer below not checking for them?

Reply:
I think this is incorrect. The sniffer issues a call to `_get_header`, which in turn tests these are present in the hdf5 file.

Reply:
Ah yep you're right here. Just did not see the keys there, got it now.